python/GafferTest/MessageHandlerTest.py

```diff
@@ -0,0 +1,90 @@
+#  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+#  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+#  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+#  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+#  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+#  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+#  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#
+##########################################################################
+
+import IECore
+
+import Gaffer
+import GafferTest
+
+class MessageHandlerTest( GafferTest.TestCase ) :
+
+	def testMessageOutSideProcessIsForwaredUnmodified( self ) :
+
+		capturingMessageHandler = IECore.CapturingMessageHandler()
+		messageHandler = Gaffer.MessageHandler( capturingMessageHandler )
+
+		# if we log a message outside a compute or hash Process then we only get the original message
+		messageHandler.handle(IECore.MessageHandler.Level.Debug, "sending out an SOS", "message in a bottle")
+
+		self.assertEqual(len( capturingMessageHandler.messages ), 1 )
+		self.assertEqual(capturingMessageHandler.messages[0].level, IECore.MessageHandler.Level.Debug)
+		self.assertEqual(capturingMessageHandler.messages[0].context, "sending out an SOS")
+		self.assertEqual(capturingMessageHandler.messages[0].message, "message in a bottle")
+
+	def testMessageInProcessGetExtraDebugInfo( self ) :
+
+		oldHandler = IECore.MessageHandler.getDefaultHandler()
+
+		capturingMessageHandler = IECore.CapturingMessageHandler()
+		messageHandler = Gaffer.MessageHandler( capturingMessageHandler )
+		IECore.MessageHandler.setDefaultHandler( messageHandler )
+
+		scriptNode = Gaffer.ScriptNode()
+
+		expression = Gaffer.Expression( "Expression" )
+
+		node = Gaffer.Node( "Node" )
+
+		node["user"].addChild( Gaffer.IntPlug( "test", defaultValue = 0, flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
+
+		scriptNode.addChild(expression)
+		scriptNode.addChild(node)
+
+		expression.setExpression( 'import IECore\nIECore.MessageHandler.output(IECore.MessageHandler.Level.Error, "testA", "testB") \n\nparent["Node"]["user"]["test"] = 1', 'python' )
+
+		node['user']['test'].getValue()
+
+		self.assertEqual(len( capturingMessageHandler.messages ), 2 )
+
+		self.assertEqual(capturingMessageHandler.messages[0].level, IECore.MessageHandler.Level.Error)
+		self.assertEqual(capturingMessageHandler.messages[0].context, "testA")
+		self.assertEqual(capturingMessageHandler.messages[0].message, "testB")
+
+		self.assertEqual(capturingMessageHandler.messages[1].level, IECore.MessageHandler.Level.Debug)
+		self.assertEqual(capturingMessageHandler.messages[1].context, "Gaffer::Process")
+		self.assertEqual(capturingMessageHandler.messages[1].message, "[ plug: 'ScriptNode.Expression.__execute', path: '', time: 1]\n")
+
+		IECore.MessageHandler.setDefaultHandler( oldHandler )
```

Comment:
We won't get here if anything above fails, and we'll be leaving the wrong handler in place. Better to leave the default handler alone and just scope the handler on the current thread for the duration of our compute :

```
with messageHandler :
    node["user"]["test"].getValue()
```